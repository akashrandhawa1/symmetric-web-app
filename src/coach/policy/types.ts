/**
 * Policy types for state-aware voice coaching system.
 *
 * This module defines the core types for a state-aware, objective-driven coach
 * that adapts its responses based on app surface, experience level, and workout context.
 */

/**
 * App surfaces representing different states in the workout flow.
 * Each surface has distinct coaching needs and policies.
 */
export type AppSurface =
  | 'home'           // User is on home screen, not actively training
  | 'pre_session'    // About to start a session, planning phase
  | 'warmup'         // Warming up, focus on form and preparation
  | 'working_set'    // Main working sets, balanced intensity
  | 'top_set'        // Peak intensity sets, pushing limits
  | 'backoff'        // Post-peak sets with reduced load
  | 'rest_overlay'   // Resting between sets
  | 'cooldown'       // Winding down the session
  | 'post_session'   // Session complete, reviewing results
  | 'recovery_mode'; // Active recovery day

/**
 * User experience bands for adaptive verbosity and topic selection.
 */
export type ExperienceBand = 'novice' | 'intermediate' | 'advanced';

/**
 * Coaching topics that can be addressed in responses.
 * Each topic aligns with specific training concepts.
 */
export type Topic =
  | 'load'              // Weight/resistance adjustments
  | 'reps'              // Repetition targets and counting
  | 'tempo'             // Movement speed and timing
  | 'depth'             // Range of motion (especially squat depth)
  | 'stance'            // Foot position and setup
  | 'symmetry'          // Left/right balance
  | 'plan'              // Session structure and strategy
  | 'readiness_budget'  // Fatigue management and remaining capacity
  | 'motivation';       // Mental engagement (use sparingly)

/**
 * Coaching objectives that drive the response strategy.
 * Each objective focuses the coach's decision-making.
 */
export type CoachObjective =
  | 'decide_next_block'    // Help user choose what to do next
  | 'execute_reps'         // Guide through current set execution
  | 'push_or_hold'         // Decide whether to increase intensity or maintain
  | 'fix_single_fault'     // Address one specific form issue
  | 'protect_budget'       // Prevent overtraining, manage fatigue
  | 'wrap_and_transition'; // End current phase and move to next

/**
 * Complete snapshot of coaching context at a given moment.
 * This is passed to the router to generate appropriate responses.
 */
export type CoachSnapshot = {
  appSurface: AppSurface;
  experienceBand: ExperienceBand;
  readiness_now: number;          // Current readiness score (0-100)
  readiness_target: number;       // Target post-session readiness (typically ~50)
  requiresChange: boolean;        // Engine signals a change is needed
  phase: 'planning' | 'executing' | 'resting';

  // Optional set data from most recent performance
  last_set?: {
    exercise: string;
    weight_lb: number;
    reps: number;
    tempo?: string;                // e.g., "2010" (2s eccentric, 0s bottom, 1s concentric, 0s top)
    depth?: 'above' | 'parallel' | 'below';
    bar_speed?: 'slow' | 'stable' | 'fast';
  };

  // Symmetry analysis
  symmetry?: {
    left_pct: number;
    right_pct: number;
  };

  // Session management
  time_left_min?: number;

  // Safety flags
  safety?: {
    pain_flag: boolean;
  };

  // User intent classification
  intent: 'report' | 'ask' | 'struggle' | 'brag' | 'stall';
  utterance: string;              // Raw user input
};

/**
 * Policy configuration that controls coaching behavior.
 * Policies are determined by app surface and modified by experience band.
 */
export type Policy = {
  objective: CoachObjective;
  allowed: Topic[];                       // Topics the coach can discuss
  banned?: Topic[];                       // Topics to explicitly avoid
  wordBudget: number;                     // Maximum words in response
  sayNothingUnlessChange?: boolean;       // Stay silent unless requiresChange is true
};

/**
 * Structured LLM response with all components of a coaching cue.
 * Format: HOOK + WHY + ACTION
 */
export type LLMReply = {
  silent?: boolean;                       // If true, coach should not speak
  hook?: string;                          // Opening statement to grab attention (1 sentence)
  why?: string;                           // Brief reasoning or context (1 clause)
  action?: string;                        // Concrete directive for user (1 instruction)
  action_type?: Topic;                    // Which topic this action addresses
  prosody?: {                            // Voice delivery characteristics
    pace?: 'slow' | 'medium' | 'fast';
    energy?: 'calm' | 'confident' | 'high';
  };
};
