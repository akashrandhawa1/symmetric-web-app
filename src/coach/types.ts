import type { WhatIfQualPayload } from './types.whatif';
export type PlanMode = 'TRAIN' | 'ACTIVE_RECOVERY' | 'FULL_REST';

export type SessionState =
  | 'READY_NOW'
  | 'POST_SESSION'
  | 'COOLDOWN_24H'
  | 'RECOVERING'
  | 'PRE_SESSION'
  | 'IN_SESSION'
  | 'POST_SET'
  | 'IDLE'
  | string;

export type Gates = { status: 'GO' | 'CAUTION' | 'STOP'; reason: string };

export type SessionLog = {
  startedAt: string;
  endedAt?: string;
  exerciseBlocks: Array<{
    name: string;
    side?: 'L' | 'R';
    sets: Array<{
      setIndex: number;
      reps: number;
      repBars?: number[];
      emg?: { rms?: number[]; ror?: number[]; symmetryPct?: number[] };
      peakRMS?: number;
      peakRoR?: number;
      rpe?: number;
      restSec?: number;
      flags?: ('failed' | 'cut_early' | 'form_decline')[];
    }>;
  }>;
  summaries?: {
    bestSetRMS?: number;
    bestSetRoR?: number;
    avgSymmetryPct?: number;
    fatigue?: { rmsDropPct: number; rorDropPct: number };
    totalSets?: number;
    totalReps?: number;
  };
};

export type HistoryAgg = {
  mvcRMSBaseline?: number;
  bestRoRBaseline?: number;
  weeklyStrengthSetsDone: number;
  weeklyStrengthTarget: number;
  lastSessionEnd?: string;
  lastSameMuscleSessionEnd?: string;
  hoursSinceLastSessionEnd?: number;
  hoursSinceLastSameMuscle?: number;
  sym7dAvg?: number;
};

export type RecoveryState = {
  readiness: number;
  readinessSlopePerHr?: number;
  soreness?: number;
  sleepHrs?: number;
  hrWarning?: boolean;
};

export type StrengthContext = {
  nowISO: string;
  muscleGroup: string;
  todaySession?: SessionLog;
  history: HistoryAgg;
  recovery: RecoveryState;
  gates: Gates;
  symmetryIdeal: number;
  strengthWindowReps: [number, number];
  fatigueZones: { rms: [number, number, number]; ror: [number, number, number] };
  didExerciseToday: boolean;
  justExercised: boolean;
  inRestWindow: boolean;
  targetReadinessMin: number;
  targetReadinessMax: number;
  postSessionCooldownHrs: number;
  fullCooldownHrs: number;
  sessionState: SessionState;
};

export type StrengthPlan = {
  mode: PlanMode;
  reasonCodes: string[];
  primaryActions: string[];
  guardrails: string[];
  nextCheck: string;
  projections?: {
    trainBlock?: { strengthGainPct: number; readinessDelta: number; expectedSets: number };
    activeRecovery30m?: { readinessDelta: number; sorenessDelta?: number };
    fullRestTonight?: { readinessDelta: number };
  };
  confidence?: number;
};

export type DecisionOption = {
  id: PlanMode;
  strength_gain_pct: number;
  readiness_delta: number;
  hours_to_recover: number;
  utility: number;
};

export type DecisionTrace = {
  options: DecisionOption[];
  chosenId: PlanMode;
  reasonCodes: string[];
};

export type CoachSummary = {
  message: string;
  cta: string;
  secondary?: string;
  what_if?: WhatIfQualPayload | null;
};

export type StrengthPlanResult = {
  plan: StrengthPlan;
  trace: DecisionTrace;
};

export type NarratorInput = {
  plan: StrengthPlan;
  receipts: {
    readiness: number;
    fatigue: { rms?: number; ror?: number };
    symmetry?: number;
    sinceSameMuscleHrs?: number;
    roiLine?: string;
  };
  decisionTrace: DecisionTrace;
};

// ============================================================================
// HOME COACH (LLM-First + App-Verified) TYPES
// ============================================================================

export type CoachMode = "TRAIN" | "ACTIVE_RECOVERY" | "FULL_REST";

export type FatigueZone = "GREEN" | "YELLOW" | "ORANGE" | "RED";

export type ReadinessBand = "HIGH" | "MID" | "LOW";

/**
 * Context payload for LLM tool: get_context()
 */
export type ContextPayload = {
  nowISO: string;
  sessionState: SessionState;
  readiness_local: number;
  symmetryPct: number | null;
  fatigue: {
    rmsDropPct: number | null;
    rorDropPct: number | null;
  };
  hoursSinceLastSameMuscle: number | null;
  weekly: {
    done: number;
    target: number;
  };
  flags: {
    hrWarning: boolean;
    sorenessHigh: boolean;
  };
  lastEndZone: FatigueZone | null;
  policy: {
    strengthWindowReps: [number, number];
    symmetryIdeal: number;
    fatigueZones: {
      rms: [number, number, number];
      ror: [number, number, number];
    };
  };
  allowed_actions: (ActionId | string)[];
};

export type ActionId = "start_strength_block" | "start_recovery_30m" | "plan_tomorrow";

/**
 * Projection for LLM tool: project_action()
 */
export type Projection = {
  action_id: ActionId;
  effects: {
    strength_gain_pct: number | null;
    readiness_delta_pts: number | null;
    recovery_hours: number | null;
  };
  summary: string;
};

/**
 * Verification result from LLM tool: verify_plan()
 */
export type VerifyResult = {
  ok: boolean;
  safe_mode: CoachMode;
  reason?: string;
};

/**
 * LLM output: Suggestion
 */
export type SuggestionJSON = {
  type: "suggestion";
  mode: CoachMode;
  message: string;
  cta: string;
  secondary?: string;
  what_if?: WhatIfQualPayload | null;
};

/**
 * LLM output: Question (clarification needed)
 */
export type QuestionJSON = {
  type: "question";
  message: string;
};

/**
 * Combined LLM response type
 */
export type CoachJSON = SuggestionJSON | QuestionJSON;

/**
 * Function calling tool definition
 */
export type ToolCall = {
  name: string;
  arguments: Record<string, unknown>;
};

export type ToolResponse = {
  name: string;
  content: unknown;
};
