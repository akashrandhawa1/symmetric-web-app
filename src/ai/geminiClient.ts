type CallArgs = {
  model: string;
  system: string;
  prompt: string;
  maxOutputTokens?: number;
  temperature?: number;
  jsonSchema?: object;
};

export async function callGemini(_args: CallArgs): Promise<string> {
  // TODO: integrate real SDK. For now, assume a fetch-like interface.
  // Return raw model text (JSON string).
  throw new Error("Integrate Gemini SDK here");
}

export type LiveSession = {
  send: (text: string) => void;
  onMessage: (cb: (deltaText: string) => void) => void;
  interrupt: () => void;
  close: () => Promise<void>;
};

export async function openGeminiLiveSession(_model: string): Promise<LiveSession> {
  // TODO: integrate real live/WS client.
  throw new Error("Integrate Gemini Live SDK here");
}
