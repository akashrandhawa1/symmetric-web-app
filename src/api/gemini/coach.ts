/**
 * API endpoint stub for Gemini Coach.
 *
 * This is a mock/stub implementation that returns canned responses when MOCK_COACH=1.
 * TODO: Wire this to the actual Gemini API endpoint for production use.
 *
 * In a real Vite app, this would be implemented as a server-side endpoint using:
 * - Vite's proxy configuration (vite.config.ts)
 * - Or a separate Express/Fastify server
 * - Or Next.js API routes if migrating to Next
 *
 * @module api/gemini/coach
 */

import type { CoachRequest, CoachAdvice } from '@/lib/coach/types';
import { inWindowResponse } from '@/mocks/coachFixtures';

/**
 * Mock handler that returns canned valid responses.
 * Set MOCK_COACH=1 to use this instead of hitting the real API.
 */
export async function handleCoachRequest(
  systemPrompt: string,
  request: CoachRequest
): Promise<CoachAdvice> {
  // In mock mode, return a canned response
  if (process.env.MOCK_COACH === '1') {
    console.log('[Mock] Returning canned coach response');
    // Simulate network delay
    await new Promise((resolve) => setTimeout(resolve, 800));
    return inWindowResponse;
  }

  // TODO: Implement real Gemini API call here
  // Example implementation structure:
  //
  // const response = await fetch('https://generativelanguage.googleapis.com/v1beta/...', {
  //   method: 'POST',
  //   headers: {
  //     'Content-Type': 'application/json',
  //     'x-goog-api-key': process.env.GEMINI_API_KEY || '',
  //   },
  //   body: JSON.stringify({
  //     contents: [{
  //       parts: [{
  //         text: `${systemPrompt}\n\nUser request:\n${JSON.stringify(request, null, 2)}`
  //       }]
  //     }],
  //     generationConfig: {
  //       temperature: 0.7,
  //       topP: 0.9,
  //       responseSchema: { ... }, // Pass Zod schema as JSON Schema
  //     }
  //   })
  // });
  //
  // const data = await response.json();
  // return data.candidates[0].content.parts[0].text; // Parse JSON from text

  throw new Error(
    'Real Gemini API not implemented yet. Set MOCK_COACH=1 to use mock responses.'
  );
}

/**
 * Express-style route handler (for reference if using server framework).
 * This would be mounted at POST /api/gemini/coach
 */
export async function POST(req: { body: { systemPrompt: string; request: CoachRequest } }) {
  try {
    const { systemPrompt, request } = req.body;

    if (!systemPrompt || !request) {
      return {
        status: 400,
        body: { error: 'Missing systemPrompt or request in body' },
      };
    }

    const advice = await handleCoachRequest(systemPrompt, request);

    return {
      status: 200,
      body: advice,
    };
  } catch (error) {
    console.error('Error in coach API:', error);
    return {
      status: 500,
      body: { error: 'Internal server error' },
    };
  }
}
