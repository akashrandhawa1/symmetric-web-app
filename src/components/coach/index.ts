/**
 * Barrel export for Rest Screen Coach components.
 * Provides clean import paths for consumer code.
 */

export { RestCoach } from './RestCoach';
export type { RestCoachProps } from './RestCoach';

export { useCountdown } from './useCountdown';
export type { UseCountdownReturn } from './useCountdown';
