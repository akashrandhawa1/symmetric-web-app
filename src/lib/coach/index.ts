/**
 * Barrel export for Rest Screen Coach library modules.
 * Provides clean import paths for types, client, and telemetry.
 */

// Types
export type {
  CoachRequest,
  CoachAdvice,
  TrainingAge,
  TargetRepRange,
  UserPrevDecision,
  PainFlag,
  AdviceType,
  EffortDelta,
  ReasonCode,
} from './types';

export { zCoachAdvice, trimToTwoLines, validateSafetyConstraints, REASON_CODE_LABELS } from './types';

// Client
export { getCoachAdvice, createMockAdvice, FALLBACK_ADVICE, SYSTEM_PROMPT } from './geminiClient';

// Telemetry
export type {
  UserDecision,
  CoachAdviceShownPayload,
  CoachUserDecisionPayload,
  CoachNextSetOutcomePayload,
} from './telemetry';

export {
  emitAdviceShown,
  emitUserDecision,
  emitNextSetOutcome,
  installTelemetryHandler,
  clearTelemetryHandler,
} from './telemetry';
