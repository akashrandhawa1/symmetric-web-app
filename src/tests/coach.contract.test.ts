/**
 * Contract tests for Rest Screen Coach.
 *
 * Tests verify:
 * 1. Schema validation (passes on valid, rejects on malformed)
 * 2. Two-line text constraint enforcement (truncation)
 * 3. Safety invariants (pain/low-signal → no aggressive effort)
 * 4. Projection gating (CI < 0.7 → hidden)
 * 5. Telemetry emission (correct payloads)
 *
 * @module tests/coach.contract.test
 */

import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import type { CoachAdvice, CoachRequest } from '@/lib/coach/types';
import { zCoachAdvice, trimToTwoLines, validateSafetyConstraints } from '@/lib/coach/types';
import { createMockAdvice, FALLBACK_ADVICE } from '@/lib/coach/geminiClient';
import {
  emitAdviceShown,
  emitUserDecision,
  emitNextSetOutcome,
  installTelemetryHandler,
  clearTelemetryHandler,
} from '@/lib/coach/telemetry';
import {
  earlyFatigueResponse,
  noFatigueResponse,
  inWindowResponse,
  painFlagResponse,
  lowSignalResponse,
  projectionAbsentResponse,
  allFixtures,
} from '@/mocks/coachFixtures';

// ============================================================================
// SCHEMA VALIDATION TESTS
// ============================================================================

describe('Schema Validation', () => {
  it('should accept valid advice from fixtures', () => {
    allFixtures.forEach(({ name, response }) => {
      const result = zCoachAdvice.safeParse(response);
      expect(result.success, `Fixture "${name}" should pass validation`).toBe(true);
    });
  });

  it('should reject advice with missing required fields', () => {
    const invalid = {
      advice_id: 'test-001',
      // Missing advice_type
      primary_text: 'Test',
      secondary_text: 'Test',
    };

    const result = zCoachAdvice.safeParse(invalid);
    expect(result.success).toBe(false);
  });

  it('should reject advice with text exceeding 140 characters', () => {
    const invalid = createMockAdvice({
      primary_text: 'A'.repeat(141), // Too long
    });

    const result = zCoachAdvice.safeParse(invalid);
    expect(result.success).toBe(false);
  });

  it('should reject advice with confidence out of bounds', () => {
    const invalid = {
      ...createMockAdvice(),
      confidence: 1.5, // > 1
    };

    const result = zCoachAdvice.safeParse(invalid);
    expect(result.success).toBe(false);
  });

  it('should reject advice with projection CI out of bounds', () => {
    const invalid = {
      ...createMockAdvice(),
      projection: {
        delta_hit_rate_pct: 10,
        ci: 1.2, // > 1
      },
    };

    const result = zCoachAdvice.safeParse(invalid);
    expect(result.success).toBe(false);
  });

  it('should accept advice with null optional fields', () => {
    const valid = createMockAdvice({
      rest_seconds: null,
      effort_delta: null,
      add_reps: null,
      swap_candidate: null,
      projection: null,
    });

    const result = zCoachAdvice.safeParse(valid);
    expect(result.success).toBe(true);
  });
});

// ============================================================================
// TWO-LINE CONSTRAINT TESTS
// ============================================================================

describe('Two-Line Text Constraint', () => {
  it('should not modify text that is within limit', () => {
    const advice = createMockAdvice({
      primary_text: 'Short text.',
      secondary_text: 'Also short.',
    });

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.primary_text).toBe(advice.primary_text);
    expect(trimmed.secondary_text).toBe(advice.secondary_text);
  });

  it('should truncate primary text exceeding 140 chars with ellipsis', () => {
    const longText = 'A'.repeat(150);
    const advice = createMockAdvice({ primary_text: longText });

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.primary_text.length).toBe(140);
    expect(trimmed.primary_text.endsWith('…')).toBe(true);
  });

  it('should truncate secondary text exceeding 140 chars with ellipsis', () => {
    const longText = 'B'.repeat(200);
    const advice = createMockAdvice({ secondary_text: longText });

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.secondary_text.length).toBe(140);
    expect(trimmed.secondary_text.endsWith('…')).toBe(true);
  });

  it('should truncate both lines if both exceed limit', () => {
    const advice = createMockAdvice({
      primary_text: 'P'.repeat(160),
      secondary_text: 'S'.repeat(180),
    });

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.primary_text.length).toBe(140);
    expect(trimmed.secondary_text.length).toBe(140);
  });

  it('should handle text at exactly 140 chars without truncation', () => {
    const exactText = 'X'.repeat(140);
    const advice = createMockAdvice({ primary_text: exactText });

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.primary_text).toBe(exactText);
    expect(trimmed.primary_text.endsWith('…')).toBe(false);
  });
});

// ============================================================================
// SAFETY INVARIANT TESTS
// ============================================================================

describe('Safety Constraints', () => {
  it('should reject aggressive effort when pain_flag > 0', () => {
    const advice = createMockAdvice({ effort_delta: 1 });
    const telemetry: CoachRequest['set_telemetry'] = {
      set_index: 1,
      rep_count: 8,
      fatigue_rep: 7,
      symmetry_pct: 90,
      rms_drop_pct: 12,
      ror_ok: true,
      signal_confidence: 0.9,
      pain_flag: 3, // Pain present
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(advice, telemetry);
    expect(isValid).toBe(false);
  });

  it('should reject aggressive effort when signal_confidence < 0.7', () => {
    const advice = createMockAdvice({ effort_delta: 1 });
    const telemetry: CoachRequest['set_telemetry'] = {
      set_index: 1,
      rep_count: 8,
      fatigue_rep: 7,
      symmetry_pct: 90,
      rms_drop_pct: 12,
      ror_ok: true,
      signal_confidence: 0.65, // Low signal
      pain_flag: 0,
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(advice, telemetry);
    expect(isValid).toBe(false);
  });

  it('should allow aggressive effort when pain=0 and signal_confidence >= 0.7', () => {
    const advice = createMockAdvice({ effort_delta: 1 });
    const telemetry: CoachRequest['set_telemetry'] = {
      set_index: 1,
      rep_count: 8,
      fatigue_rep: 7,
      symmetry_pct: 90,
      rms_drop_pct: 12,
      ror_ok: true,
      signal_confidence: 0.92,
      pain_flag: 0,
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(advice, telemetry);
    expect(isValid).toBe(true);
  });

  it('should allow non-aggressive advice even with pain', () => {
    const advice = createMockAdvice({ effort_delta: 0 });
    const telemetry: CoachRequest['set_telemetry'] = {
      set_index: 1,
      rep_count: 8,
      fatigue_rep: 7,
      symmetry_pct: 90,
      rms_drop_pct: 12,
      ror_ok: true,
      signal_confidence: 0.9,
      pain_flag: 5,
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(advice, telemetry);
    expect(isValid).toBe(true);
  });

  it('should allow effort reduction with pain or low signal', () => {
    const advice = createMockAdvice({ effort_delta: -1 });
    const telemetry: CoachRequest['set_telemetry'] = {
      set_index: 1,
      rep_count: 8,
      fatigue_rep: 7,
      symmetry_pct: 90,
      rms_drop_pct: 12,
      ror_ok: true,
      signal_confidence: 0.6,
      pain_flag: 2,
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(advice, telemetry);
    expect(isValid).toBe(true);
  });
});

// ============================================================================
// PROJECTION GATING TESTS
// ============================================================================

describe('Projection Gating', () => {
  it('should show projection when CI >= 0.7', () => {
    const advice = createMockAdvice({
      projection: {
        delta_hit_rate_pct: 15,
        ci: 0.85,
      },
    });

    expect(advice.projection).not.toBeNull();
    expect(advice.projection!.ci).toBeGreaterThanOrEqual(0.7);
  });

  it('should hide projection in UI when CI < 0.7', () => {
    const advice = projectionAbsentResponse;

    expect(advice.projection).not.toBeNull();
    expect(advice.projection!.ci).toBeLessThan(0.7);

    // UI logic: should not display
    const shouldDisplay = advice.projection && advice.projection.ci >= 0.7;
    expect(shouldDisplay).toBe(false);
  });

  it('should handle null projection gracefully', () => {
    const advice = createMockAdvice({ projection: null });

    const shouldDisplay = !!(advice.projection && advice.projection.ci >= 0.7);
    expect(shouldDisplay).toBe(false);
  });
});

// ============================================================================
// TELEMETRY TESTS
// ============================================================================

describe('Telemetry Events', () => {
  const mockEmit = vi.fn();

  beforeEach(() => {
    mockEmit.mockClear();
    installTelemetryHandler(mockEmit);
  });

  afterEach(() => {
    clearTelemetryHandler();
  });

  it('should emit coach_advice_shown with correct payload', () => {
    const advice = createMockAdvice({
      advice_id: 'test-advice-001',
      advice_type: 'rest',
      confidence: 0.88,
      projection: { delta_hit_rate_pct: 10, ci: 0.75 },
      telemetry_tags: ['test', 'rest'],
    });

    emitAdviceShown(advice);

    expect(mockEmit).toHaveBeenCalledTimes(1);
    expect(mockEmit).toHaveBeenCalledWith('coach_advice_shown', {
      advice_id: 'test-advice-001',
      advice_type: 'rest',
      projection_ci: 0.75,
      confidence: 0.88,
      telemetry_tags: ['test', 'rest'],
      timestamp: expect.any(Number),
    });
  });

  it('should emit coach_user_decision with correct payload', () => {
    emitUserDecision({
      advice_id: 'test-advice-002',
      decision: 'skip',
      reason_code: 'short_on_time',
      time_to_decision_ms: 5432,
    });

    expect(mockEmit).toHaveBeenCalledTimes(1);
    expect(mockEmit).toHaveBeenCalledWith('coach_user_decision', {
      advice_id: 'test-advice-002',
      decision: 'skip',
      reason_code: 'short_on_time',
      time_to_decision_ms: 5432,
      timestamp: expect.any(Number),
    });
  });

  it('should emit coach_next_set_outcome with correct payload', () => {
    emitNextSetOutcome({
      advice_id: 'test-advice-003',
      actual_reps: 9,
      actual_fatigue_rep: 8,
      actual_symmetry_pct: 93,
      advice_followed: true,
    });

    expect(mockEmit).toHaveBeenCalledTimes(1);
    expect(mockEmit).toHaveBeenCalledWith('coach_next_set_outcome', {
      advice_id: 'test-advice-003',
      actual_reps: 9,
      actual_fatigue_rep: 8,
      actual_symmetry_pct: 93,
      advice_followed: true,
      timestamp: expect.any(Number),
    });
  });

  it('should handle missing telemetry handler gracefully', () => {
    clearTelemetryHandler();

    expect(() => {
      emitAdviceShown(createMockAdvice());
    }).not.toThrow();
  });
});

// ============================================================================
// FALLBACK ADVICE TESTS
// ============================================================================

describe('Fallback Advice', () => {
  it('should have valid fallback advice structure', () => {
    const result = zCoachAdvice.safeParse(FALLBACK_ADVICE);
    expect(result.success).toBe(true);
  });

  it('should prioritize safety in fallback', () => {
    expect(FALLBACK_ADVICE.advice_type).toBe('check_signal');
    expect(FALLBACK_ADVICE.safety.suppress_load_calls).toBe(true);
    expect(FALLBACK_ADVICE.effort_delta).toBe(0);
  });

  it('should provide conservative rest period in fallback', () => {
    expect(FALLBACK_ADVICE.rest_seconds).toBe(90);
  });
});

// ============================================================================
// FIXTURE SNAPSHOT TESTS
// ============================================================================

describe('Fixture Snapshots', () => {
  it('should match snapshot for early fatigue scenario', () => {
    expect(earlyFatigueResponse).toMatchSnapshot();
  });

  it('should match snapshot for no fatigue scenario', () => {
    expect(noFatigueResponse).toMatchSnapshot();
  });

  it('should match snapshot for in-window scenario', () => {
    expect(inWindowResponse).toMatchSnapshot();
  });

  it('should match snapshot for pain flag scenario', () => {
    expect(painFlagResponse).toMatchSnapshot();
  });

  it('should match snapshot for low signal scenario', () => {
    expect(lowSignalResponse).toMatchSnapshot();
  });

  it('all fixtures should be valid', () => {
    allFixtures.forEach(({ name, request, response }) => {
      // Validate request structure
      expect(request.version).toBe('coach-v1');
      expect(request.user_profile).toBeDefined();
      expect(request.session_context).toBeDefined();
      expect(request.set_telemetry).toBeDefined();
      expect(request.ui_capabilities).toBeDefined();

      // Validate response schema
      const result = zCoachAdvice.safeParse(response);
      expect(result.success, `Fixture "${name}" response should be valid`).toBe(true);
    });
  });
});

// ============================================================================
// INTEGRATION TESTS
// ============================================================================

describe('Integration Scenarios', () => {
  it('should enforce safety when both pain and low signal present', () => {
    const dangerousAdvice = createMockAdvice({
      effort_delta: 1,
      add_reps: 3,
    });

    const dangerousTelemetry: CoachRequest['set_telemetry'] = {
      set_index: 2,
      rep_count: 7,
      fatigue_rep: null,
      symmetry_pct: 88,
      rms_drop_pct: 10,
      ror_ok: true,
      signal_confidence: 0.58,
      pain_flag: 4,
      user_prev_decision: 'none',
    };

    const isValid = validateSafetyConstraints(dangerousAdvice, dangerousTelemetry);
    expect(isValid).toBe(false);
  });

  it('should allow progressive overload in safe conditions', () => {
    const progressiveAdvice = createMockAdvice({
      effort_delta: 1,
      add_reps: 2,
    });

    const safeTelemetry: CoachRequest['set_telemetry'] = {
      set_index: 3,
      rep_count: 10,
      fatigue_rep: null,
      symmetry_pct: 94,
      rms_drop_pct: 8,
      ror_ok: true,
      signal_confidence: 0.96,
      pain_flag: 0,
      user_prev_decision: 'did',
    };

    const isValid = validateSafetyConstraints(progressiveAdvice, safeTelemetry);
    expect(isValid).toBe(true);
  });

  it('should handle edge case: exactly 140 chars after model response', () => {
    const advice = createMockAdvice({
      primary_text: 'X'.repeat(140),
      secondary_text: 'Y'.repeat(140),
    });

    const parseResult = zCoachAdvice.safeParse(advice);
    expect(parseResult.success).toBe(true);

    const trimmed = trimToTwoLines(advice);
    expect(trimmed.primary_text.length).toBe(140);
    expect(trimmed.secondary_text.length).toBe(140);
  });
});
