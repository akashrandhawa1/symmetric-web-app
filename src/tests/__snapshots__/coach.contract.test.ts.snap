// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fixture Snapshots > should match snapshot for early fatigue scenario 1`] = `
{
  "add_reps": null,
  "advice_id": "adv-001-early-fatigue",
  "advice_type": "effort_change",
  "ask_reason_on_skip_or_override": true,
  "confidence": 0.88,
  "effort_delta": -1,
  "internal_rationale": "Fatigue rep 4 is well below target window 7-10; reduce load.",
  "offer_swap": false,
  "primary_text": "Fatigue crashed the party at rep 4—tomorrow's legs would riot.",
  "projection": {
    "ci": 0.82,
    "delta_hit_rate_pct": 12,
  },
  "rest_seconds": 100,
  "safety": {
    "end_exercise": false,
    "suppress_load_calls": false,
  },
  "secondary_text": "Ease the load about 5% and rest 100s so recovery stays on your side.",
  "swap_candidate": null,
  "telemetry_tags": [
    "early_fatigue",
    "effort_down",
  ],
}
`;

exports[`Fixture Snapshots > should match snapshot for in-window scenario 1`] = `
{
  "add_reps": null,
  "advice_id": "adv-003-in-window",
  "advice_type": "rest",
  "ask_reason_on_skip_or_override": false,
  "confidence": 0.94,
  "effort_delta": 0,
  "offer_swap": false,
  "primary_text": "Fatigue landed right at rep 8—money time.",
  "projection": null,
  "rest_seconds": 90,
  "safety": {
    "end_exercise": false,
    "suppress_load_calls": false,
  },
  "secondary_text": "Hold this plan and rest 90s to keep stacking reliable strength.",
  "swap_candidate": null,
  "telemetry_tags": [
    "in_window",
    "maintain",
  ],
}
`;

exports[`Fixture Snapshots > should match snapshot for low signal scenario 1`] = `
{
  "add_reps": null,
  "advice_id": "adv-006-low-signal",
  "advice_type": "check_signal",
  "ask_reason_on_skip_or_override": false,
  "confidence": 0.68,
  "effort_delta": 0,
  "offer_swap": false,
  "primary_text": "Signal quality slipped to 62%, so the data's sketchy.",
  "projection": null,
  "rest_seconds": 90,
  "safety": {
    "end_exercise": false,
    "suppress_load_calls": true,
  },
  "secondary_text": "Reset the electrode, rest 90s, and retry at the same effort for clean guidance.",
  "swap_candidate": null,
  "telemetry_tags": [
    "low_signal",
    "check_signal",
  ],
}
`;

exports[`Fixture Snapshots > should match snapshot for no fatigue scenario 1`] = `
{
  "add_reps": 2,
  "advice_id": "adv-002-no-fatigue",
  "advice_type": "effort_change",
  "ask_reason_on_skip_or_override": false,
  "confidence": 0.91,
  "effort_delta": 1,
  "offer_swap": false,
  "primary_text": "You breezed past the fatigue window—readiness is screaming green.",
  "projection": {
    "ci": 0.89,
    "delta_hit_rate_pct": 18,
  },
  "rest_seconds": 90,
  "safety": {
    "end_exercise": false,
    "suppress_load_calls": false,
  },
  "secondary_text": "Add two reps or bump effort +1, then rest 90s to turn that headroom into strength.",
  "swap_candidate": null,
  "telemetry_tags": [
    "no_fatigue",
    "effort_up",
  ],
}
`;

exports[`Fixture Snapshots > should match snapshot for pain flag scenario 1`] = `
{
  "add_reps": null,
  "advice_id": "adv-005-pain",
  "advice_type": "check_signal",
  "ask_reason_on_skip_or_override": false,
  "confidence": 0.76,
  "effort_delta": 0,
  "offer_swap": true,
  "primary_text": "Pain signal lit up—no PR is worth grinding through that.",
  "projection": null,
  "rest_seconds": 90,
  "safety": {
    "end_exercise": false,
    "suppress_load_calls": true,
  },
  "secondary_text": "Take 90s, then stop or swap to leg press if it still bites so the joint can calm down.",
  "swap_candidate": "Leg press",
  "telemetry_tags": [
    "pain_flag",
    "safety_check",
  ],
}
`;
